File: sdk/java/src/main/java/cloud/oresoftware/nextloggers/NextLoggers.java
package cloud.oresoftware.nextloggers;

import java.time.Instant;
import java.util.ArrayList;
import java.util.Collections;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Objects;
import java.util.UUID;
import java.util.concurrent.Callable;
import java.util.function.Consumer;

/** Dependency-free Java implementation of the next-loggers/v1 contract. */
public final class NextLoggers {
  public static final String SCHEMA = "next-loggers/v1";
  private static final ThreadLocal<Context> CONTEXT = new ThreadLocal<>();

  private NextLoggers() {}

  public enum Level {
    TRACE(1), DEBUG(5), INFO(9), WARN(13), ERROR(17), FATAL(21);

    private final int otelSeverityNumber;

    Level(int otelSeverityNumber) {
      this.otelSeverityNumber = otelSeverityNumber;
    }

    public int otelSeverityNumber() {
      return otelSeverityNumber;
    }
  }

  public record Context(
      String traceId,
      String spanId,
      int traceFlags,
      String traceState,
      Map<String, Object> fields,
      List<String> tags) {
    public Context {
      fields = immutableCopy(fields);
      tags = tags == null ? List.of() : List.copyOf(tags);
    }

    public static Context trace(String traceId, String spanId) {
      return new Context(traceId, spanId, 1, null, Map.of(), List.of("otel"));
    }
  }

  public static Context currentContext() {
    return CONTEXT.get();
  }

  /** Scoped ThreadLocal context; always restores the prior frame. */
  public static <T> T withContext(Context context, Callable<T> callback) throws Exception {
    Objects.requireNonNull(callback, "callback");
    Context previous = CONTEXT.get();
    if (context == null) {
      CONTEXT.remove();
    } else {
      CONTEXT.set(context);
    }
    try {
      return callback.call();
    } finally {
      if (previous == null) {
        CONTEXT.remove();
      } else {
        CONTEXT.set(previous);
      }
    }
  }

  @FunctionalInterface
  public interface Transport {
    void write(Map<String, Object> record) throws Exception;

    default boolean isOtel() {
      return "opentelemetry".equalsIgnoreCase(name());
    }

    default String name() {
      return "";
    }
  }

  /** Application-owned OTEL sink. No global provider or instrumentation is installed. */
  public static final class OtelTransport implements Transport {
    private final Consumer<Map<String, Object>> sink;

    public OtelTransport(Consumer<Map<String, Object>> sink) {
      this.sink = Objects.requireNonNull(sink, "sink");
    }

    @Override
    public void write(Map<String, Object> record) {
      Level level = Level.valueOf(String.valueOf(record.get("level")));
      Map<String, Object> attributes = new LinkedHashMap<>();
      attributes.put("service.name", record.get("appName"));
      attributes.put("next_logger.schema", record.get("schema"));
      attributes.put("next_logger.runtime", record.get("runtime"));
      attributes.put("log.record.uid", record.get("id"));
      copyIfPresent(record, attributes, "traceId", "trace.id");
      Object fields = record.get("fields");
      if (fields instanceof Map<?, ?> map) {
        map.forEach((key, value) -> attributes.put("next_logger.field." + key, value));
      }
      Map<String, Object> otelRecord = new LinkedHashMap<>();
      otelRecord.put("body", record.get("message"));
      otelRecord.put("severityText", level.name());
      otelRecord.put("severityNumber", level.otelSeverityNumber());
      otelRecord.put("timestamp", record.get("timestamp"));
      otelRecord.put("attributes", Collections.unmodifiableMap(attributes));
      sink.accept(Collections.unmodifiableMap(otelRecord));
    }

    @Override
    public boolean isOtel() {
      return true;
    }

    @Override
    public String name() {
      return "opentelemetry";
    }
  }

  /** Client-safe Supabase transport; the application supplies its authenticated sender. */
  public static final class SupabaseTransport implements Transport {
    private final Consumer<Map<String, Object>> sender;

    public SupabaseTransport(Consumer<Map<String, Object>> sender) {
      this.sender = Objects.requireNonNull(sender, "sender");
    }

    @Override
    public void write(Map<String, Object> record) {
      sender.accept(record);
    }
  }

  public static final class LogEvent {
    private final Logger logger;
    private final Level level;
    private final String message;
    private final Map<String, Object> fields;
    private final Context context;
    private Boolean otelEnabled;

    private LogEvent(Logger logger, Level level, String message, Map<String, Object> fields) {
      this.logger = logger;
      this.level = Objects.requireNonNull(level, "level");
      this.message = message;
      this.fields = immutableCopy(fields);
      this.context = currentContext();
    }

    public LogEvent useOtel() {
      return withOtel(true);
    }

    public LogEvent notOtel() {
      return withOtel(false);
    }

    public LogEvent withOtel(boolean enabled) {
      this.otelEnabled = enabled;
      return this;
    }

    public LogEvent resetOtel() {
      this.otelEnabled = null;
      return this;
    }

    public boolean isOtelEnabled(boolean fallback) {
      return otelEnabled == null ? fallback : otelEnabled;
    }

    public Map<String, Object> send() throws Exception {
      return logger.emitEvent(this);
    }
  }

  public static final class Logger {
    private final String appName;
    private final String name;
    private final String runtime;
    private final Map<String, Object> fields;
    private final List<Transport> transports;
    private volatile boolean otelEnabled;

    public Logger(
        String appName,
        String name,
        String runtime,
        Map<String, Object> fields,
        List<Transport> transports) {
      this(appName, name, runtime, fields, transports, true);
    }

    public Logger(
        String appName,
        String name,
        String runtime,
        Map<String, Object> fields,
        List<Transport> transports,
        boolean otelEnabled) {
      this.appName = requireText(appName, "appName");
      this.name = name;
      this.runtime = runtime == null || runtime.isBlank() ? "java" : runtime;
      this.fields = immutableCopy(fields);
      this.transports = transports == null ? List.of() : List.copyOf(transports);
      this.otelEnabled = otelEnabled;
    }

    public Logger(String appName, List<Transport> transports) {
      this(appName, null, "java", Map.of(), transports);
    }

    public Map<String, Object> log(Level level, String message, Map<String, Object> eventFields)
        throws Exception {
      return event(level, message, eventFields).send();
    }

    public LogEvent event(Level level, String message, Map<String, Object> eventFields) {
      return new LogEvent(this, level, message, eventFields);
    }

    public Logger setOtelEnabled(boolean enabled) {
      this.otelEnabled = enabled;
      return this;
    }

    public Logger useOtel() {
      return setOtelEnabled(true);
    }

    public Logger notOtel() {
      return setOtelEnabled(false);
    }

    public boolean isOtelEnabled() {
      return otelEnabled;
    }

    private Map<String, Object> emitEvent(LogEvent event) throws Exception {
      Level level = event.level;
      String message = event.message;
      Map<String, Object> eventFields = event.fields;
      Context context = event.context;
      Map<String, Object> mergedFields = new LinkedHashMap<>(fields);
      if (context != null) {
        mergedFields.putAll(context.fields());
        putIfText(mergedFields, "otel.span_id", context.spanId());
        mergedFields.put("otel.trace_flags", context.traceFlags());
        putIfText(mergedFields, "otel.trace_state", context.traceState());
      }
      if (eventFields != null) {
        mergedFields.putAll(eventFields);
      }

      Map<String, Object> record = new LinkedHashMap<>();
      record.put("schema", SCHEMA);
      record.put("id", UUID.randomUUID().toString());
      record.put("timestamp", Instant.now().toString());
      record.put("level", level.name());
      record.put("runtime", runtime);
      record.put("appName", appName);
      if (name != null && !name.isBlank()) {
        record.put("name", name);
      }
      record.put("message", message == null ? "" : message);
      record.put("values", List.of(message == null ? "" : message));
      record.put("fields", Collections.unmodifiableMap(mergedFields));
      if (context != null && context.traceId() != null && !context.traceId().isBlank()) {
        record.put("traceId", context.traceId());
        record.put("traceIds", List.of(context.traceId()));
      }
      if (context != null && !context.tags().isEmpty()) {
        record.put("tags", context.tags());
      }
      Map<String, Object> immutable = Collections.unmodifiableMap(record);
      for (Transport transport : transports) {
        if (transport.isOtel() && !event.isOtelEnabled(otelEnabled)) {
          continue;
        }
        transport.write(immutable);
      }
      return immutable;
    }

    public Map<String, Object> info(String message, Map<String, Object> fields) throws Exception {
      return log(Level.INFO, message, fields);
    }

    public Map<String, Object> error(String message, Map<String, Object> fields) throws Exception {
      return log(Level.ERROR, message, fields);
    }
  }

  public static String toJson(Object value) {
    if (value == null) return "null";
    if (value instanceof String text) return '"' + escape(text) + '"';
    if (value instanceof Number || value instanceof Boolean) return String.valueOf(value);
    if (value instanceof Map<?, ?> map) {
      List<String> entries = new ArrayList<>();
      map.forEach((key, item) -> entries.add(toJson(String.valueOf(key)) + ':' + toJson(item)));
      return "{" + String.join(",", entries) + "}";
    }
    if (value instanceof Iterable<?> values) {
      List<String> items = new ArrayList<>();
      values.forEach(item -> items.add(toJson(item)));
      return "[" + String.join(",", items) + "]";
    }
    return toJson(String.valueOf(value));
  }

  private static String escape(String value) {
    return value
        .replace("\\", "\\\\")
        .replace("\"", "\\\"")
        .replace("\n", "\\n")
        .replace("\r", "\\r")
        .replace("\t", "\\t");
  }

  private static Map<String, Object> immutableCopy(Map<String, Object> source) {
    return source == null ? Map.of() : Collections.unmodifiableMap(new LinkedHashMap<>(source));
  }

  private static String requireText(String value, String name) {
    if (value == null || value.isBlank()) throw new IllegalArgumentException(name + " is required");
    return value;
  }

  private static void putIfText(Map<String, Object> target, String key, String value) {
    if (value != null && !value.isBlank()) target.put(key, value);
  }

  private static void copyIfPresent(
      Map<String, Object> source, Map<String, Object> target, String sourceKey, String targetKey) {
    if (source.containsKey(sourceKey)) target.put(targetKey, source.get(sourceKey));
  }
}

File: sdk/ruby/lib/oresoftware/next_loggers.rb
# frozen_string_literal: true

require "securerandom"
require "time"
require_relative "next_loggers/version"

module ORESoftware
  # Dependency-free Ruby implementation of the next-loggers/v1 wire contract.
  module NextLoggers
    SCHEMA = "next-loggers/v1"
    LEVELS = %w[TRACE DEBUG INFO WARN ERROR FATAL].freeze
    OTEL_SEVERITY_NUMBERS = {
      "TRACE" => 1,
      "DEBUG" => 5,
      "INFO" => 9,
      "WARN" => 13,
      "ERROR" => 17,
      "FATAL" => 21
    }.freeze
    CONTEXT_KEY = :__oresoftware_next_loggers_context

    Context = Struct.new(
      :trace_id,
      :span_id,
      :trace_flags,
      :trace_state,
      :fields,
      :tags,
      keyword_init: true
    ) do
      def initialize(**values)
        super
        self.trace_flags = trace_flags.nil? ? 1 : Integer(trace_flags)
        self.fields = (fields || {}).each_with_object({}) do |(key, value), out|
          out[key.to_s] = value
        end.freeze
        self.tags = Array(tags).map(&:to_s).uniq.freeze
        freeze
      end
    end

    module_function

    def current_context
      Thread.current.thread_variable_get(CONTEXT_KEY)
    end

    # Install context for the current thread and always restore the prior frame.
    def with_context(context)
      previous = current_context
      Thread.current.thread_variable_set(CONTEXT_KEY, normalize_context(context))
      yield
    ensure
      Thread.current.thread_variable_set(CONTEXT_KEY, previous)
    end

    def normalize_context(context)
      return nil if context.nil?
      return context if context.is_a?(Context)
      raise ArgumentError, "context must be a Hash, Context, or nil" unless context.is_a?(Hash)

      Context.new(
        trace_id: context[:trace_id] || context["trace_id"] || context[:traceId] || context["traceId"],
        span_id: context[:span_id] || context["span_id"] || context[:spanId] || context["spanId"],
        trace_flags: context[:trace_flags] || context["trace_flags"] || context[:traceFlags] || context["traceFlags"],
        trace_state: context[:trace_state] || context["trace_state"] || context[:traceState] || context["traceState"],
        fields: context[:fields] || context["fields"],
        tags: context[:tags] || context["tags"]
      )
    end

    # Application-owned OTEL sink. No global provider or runtime patching occurs.
    class OtelTransport
      def initialize(&sink)
        raise ArgumentError, "sink block is required" unless sink

        @sink = sink
      end

      def write(record)
        attributes = {
          "service.name" => record.fetch("appName"),
          "next_logger.schema" => record.fetch("schema"),
          "next_logger.runtime" => record.fetch("runtime"),
          "log.record.uid" => record.fetch("id")
        }
        attributes["trace.id"] = record["traceId"] if record.key?("traceId")
        record.fetch("fields").each do |key, value|
          attributes["next_logger.field.#{key}"] = value
        end

        @sink.call(
          {
            "body" => record.fetch("message"),
            "severityText" => record.fetch("level"),
            "severityNumber" => OTEL_SEVERITY_NUMBERS.fetch(record.fetch("level")),
            "timestamp" => record.fetch("timestamp"),
            "attributes" => attributes.freeze
          }.freeze
        )
      end

      def otel? = true
      def name = "opentelemetry"
    end

    # Client-safe Supabase transport; the application supplies its sender.
    class SupabaseTransport
      def initialize(&sender)
        raise ArgumentError, "sender block is required" unless sender

        @sender = sender
      end

      def write(record)
        @sender.call(record)
      end
    end

    class LogEvent
      attr_reader :logger, :level, :message, :fields

      def initialize(logger, level, message, fields = {})
        @logger = logger
        @level = level
        @message = message
        @fields = fields
        @otel_enabled = nil
      end

      def use_otel = with_otel(true)
      def not_otel = with_otel(false)

      def with_otel(enabled)
        @otel_enabled = !!enabled
        self
      end

      def reset_otel
        @otel_enabled = nil
        self
      end

      def otel_enabled?(fallback)
        @otel_enabled.nil? ? !!fallback : @otel_enabled
      end

      def send
        logger.emit_event(self)
      end
    end

    class Logger
      def initialize(
        app_name:,
        name: nil,
        runtime: "ruby",
        fields: {},
        transports: [],
        otel: true,
        id_factory: -> { SecureRandom.uuid },
        clock: -> { Time.now.utc.iso8601(3) }
      )
        raise ArgumentError, "app_name is required" if app_name.to_s.strip.empty?
        raise ArgumentError, "id_factory must respond to call" unless id_factory.respond_to?(:call)
        raise ArgumentError, "clock must respond to call" unless clock.respond_to?(:call)

        @app_name = app_name.to_s.freeze
        @name = name&.to_s
        @runtime = runtime.to_s.strip.empty? ? "ruby" : runtime.to_s
        @fields = stringify_keys(fields).freeze
        @transports = Array(transports).freeze
        @otel_enabled = !!otel
        @id_factory = id_factory
        @clock = clock
      end

      def log(level, message, fields = {})
        event(level, message, fields).send
      end

      def event(level, message, fields = {})
        LogEvent.new(self, level, message, fields)
      end

      def set_otel_enabled(enabled)
        @otel_enabled = !!enabled
        self
      end

      def use_otel = set_otel_enabled(true)
      def not_otel = set_otel_enabled(false)
      def otel_enabled? = @otel_enabled

      def emit_event(event)
        level_name = event.level.to_s.upcase
        raise ArgumentError, "unsupported level: #{event.level}" unless LEVELS.include?(level_name)

        context = NextLoggers.current_context
        merged_fields = @fields.merge(context&.fields || {})
        if context
          merged_fields["otel.span_id"] = context.span_id unless blank?(context.span_id)
          merged_fields["otel.trace_flags"] = context.trace_flags
          merged_fields["otel.trace_state"] = context.trace_state unless blank?(context.trace_state)
        end
        merged_fields.merge!(stringify_keys(event.fields))
        merged_fields.freeze

        text = event.message.to_s
        record = {
          "schema" => SCHEMA,
          "id" => @id_factory.call.to_s,
          "timestamp" => @clock.call.to_s,
          "level" => level_name,
          "runtime" => @runtime,
          "appName" => @app_name,
          "message" => text,
          "values" => [text].freeze,
          "fields" => merged_fields
        }
        record["name"] = @name unless blank?(@name)
        unless context.nil? || blank?(context.trace_id)
          record["traceId"] = context.trace_id.to_s
          record["traceIds"] = [context.trace_id.to_s].freeze
        end
        record["tags"] = context.tags unless context.nil? || context.tags.empty?
        record.freeze

        @transports.each do |transport|
          name = transport.respond_to?(:name) ? transport.name.to_s.downcase : ""
          is_otel = (transport.respond_to?(:otel?) && transport.otel?) || name == "opentelemetry"
          next if is_otel && !event.otel_enabled?(@otel_enabled)

          if transport.respond_to?(:write)
            transport.write(record)
          elsif transport.respond_to?(:call)
            transport.call(record)
          else
            raise ArgumentError, "transport must respond to write or call"
          end
        end
        record
      end

      LEVELS.each do |level|
        define_method(level.downcase) do |message, fields = {}|
          log(level, message, fields)
        end
      end

      private

      def stringify_keys(value)
        raise ArgumentError, "fields must be a Hash" unless value.is_a?(Hash)

        value.each_with_object({}) { |(key, item), out| out[key.to_s] = item }
      end

      def blank?(value)
        value.nil? || value.to_s.strip.empty?
      end
    end
  end
end
